src/query/functions/temporal/base.go

```diff
@@ -0,0 +1,498 @@
+
+	if leftRangeStart.Start.Before(queryStartBounds.Start) {
+		leftRangeStart = queryStartBounds
+	}
+
+	rightRangeStart := bounds.Next(maxBlocks)
+	queryEndBounds := bounds.Nearest(c.transformOpts.TimeSpec.End.Add(-1 * bounds.StepSize))
+
+	if rightRangeStart.Start.After(queryEndBounds.Start) {
+		rightRangeStart = queryEndBounds
+	}
+
+	// Process left side of the range
+	leftBlks, emptyLeftBlocks, err := c.processLeft(bounds, maxBlocks, leftRangeStart)
+	if err != nil {
+		return err
+	}
+
+	processRequests := make([]processRequest, 0, len(leftBlks))
+	// If we have all blocks for the left range in the cache, then process the current block
+	if !emptyLeftBlocks {
+		processRequests = append(processRequests, processRequest{blk: b, deps: leftBlks, bounds: bounds})
+	}
+
+	leftBlks = append(leftBlks, b)
+
+	// Process right side of the range
+	rightBlks, emptyRightBlocks, err := c.processRight(bounds, maxBlocks, rightRangeStart)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(rightBlks); i++ {
+		lStart := maxBlocks - i
+		if lStart > len(leftBlks) {
+			continue
+		}
+
+		deps := leftBlks[len(leftBlks)-lStart:]
+		deps = append(deps, rightBlks[:i]...)
+		processRequests = append(processRequests, processRequest{blk: rightBlks[i], deps: deps, bounds: bounds.Next(i + 1)})
+	}
+
+	// If either the left range or right range wasn't fully processed then cache the current block
+	if emptyLeftBlocks || emptyRightBlocks {
+		if err := c.cache.add(bounds.Start, b); err != nil {
+			return err
+		}
+	}
+
+	return c.processCompletedBlocks(processRequests, queryStartBounds, queryEndBounds, maxBlocks)
+}
+
+// processLeft processes the current block. For the current block, figure out whether we have enough previous blocks which can help process it
+func (c *baseNode) processLeft(bounds block.Bounds, maxBlocks int, leftRangeStart block.Bounds) ([]block.Block, bool, error) {
+	numBlocks := int(bounds.Start.Sub(leftRangeStart.Start) / bounds.Duration)
+	leftBlks, err := c.cache.multiGet(leftRangeStart, numBlocks, true)
+	if err != nil {
+		return nil, false, err
+	}
+	return leftBlks, len(leftBlks) != numBlocks, nil
+}
+
+// processRight processes blocks after current block. This is done by fetching all contiguous right blocks until the right range
+func (c *baseNode) processRight(bounds block.Bounds, maxBlocks int, rightRangeStart block.Bounds) ([]block.Block, bool, error) {
+	numBlocks := int(rightRangeStart.Start.Sub(bounds.Start) / bounds.Duration)
+	rightBlks, err := c.cache.multiGet(bounds.Next(1), numBlocks, false)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return rightBlks, len(rightBlks) != maxBlocks, nil
+}
+
+// processCompletedBlocks processes all blocks for which are dependant blocks are present
+func (c *baseNode) processCompletedBlocks(processRequests []processRequest, queryStartBounds, queryEndBounds block.Bounds, maxBlocks int) error {
+	processedKeys := make([]time.Time, len(processRequests))
```

Comment:
Should this lock the mutex? Similar for processLeft and processRight

Reply:
markProcessed and sweep are the only ones mutating the blocks. markProcessed is already within a mutex and i'm not sure if sweep needs it since its fine for sweep to miss a few blocks and they can be sweeped later on. Not sure if processLeft and processRight needs a mutex since the multiGet is already locked. 

We might have to revisit this later